0000-richer-documents.rst

```diff
@@ -0,0 +1,460 @@
+  tyconAnn = ... -- wrap the type constructor in a suitable way
+
+  typeSigErrMsg :: SDoc ErrAnnotation
+  typeSigErrMsg =
+         text "Couldn't match expected type"
+     <+> embed (tyconAnn tc)
+     <+> text "with"
+     <+> embed (tyconAnn tc')
+
+Once the annotations are emitted, GHC API consumers would
+be able to get their hands on them when a compilation
+returns non-empty bags of ``ErrMsg`` or ``WarnMsg`` values, and could
+decide to use them to apply the following ideas or others in the same spirit.
+
+* A REPL front-end or IDE tool might implement color-coded output,
+  choosing a token's color by its syntactic class (e.g. type constructor,
+  data constructor, or identifier), its name or some other criterion
+  entirely.
+
+* A REPL front-end or IDE tool might allow users the ability to
+  interactively navigate a type in a type error and, for instance,
+  allow the user to interactively expand type synonyms, show kind
+  signatures, etc.
+
+* A REPL front-end or IDE tool might allow users the ability to toggle a
+  setting in order to display expressions, types and other AST related
+  entities in their AST form instead of pretty-printed. This could be useful
+  for anyone working on plugins or GHC itself.
+
+Below is a simple example of a GHC API program that loads ``M.hs``,
+collects the annotations contained in the errors and prints them (assuming
+an ``Outputable`` instance for ``ErrAnnotation`` and that we do have
+``collectAnns``).::
+
+  import Bag
+  import DynFlags
+  import GHC
+  import GHC.Paths ( libdir )
+  import HscTypes
+
+  main :: IO ()
+  main = do
+    res <- run
+    case res of
+      Right _   -> return ()
+      Left anns -> putStrLn $ "Got " ++ show (length anns) ++ " annotations"
+
+  run :: IO (Either [ErrAnnotation] SuccessFlag)
+  run = runGhc (Just libdir) $ do
+    dflags <- getSessionDynFlags
+    setSessionDynFlags dflags
+    target <- guessTarget "M.hs" Nothing
+    setTargets [target]
+    handleSourceErrors (return . Left . handleErrs dflags)
+                       (Right <$> load LoadAllTargets)
+
+  handleErrs :: DynFlags -> SourceError -> [ErrAnnotation]
+  handleErrs dflags e = concatMap (errMsgAnns dflags)
+                      $ bagToList (srcErrorMessages e)
+
+  errMsgAnns :: DynFlags -> ErrMsg -> [ErrAnnotation]
+  errMsgAnns dflags errmsg = collectAnns sdocctx sdoc
+
+    where -- we get the error document (consisting of several 'SDoc's)
+          errdoc :: ErrDoc
+          errdoc = errMsgDoc errmsg
+
+	  -- we "format" the ErrDoc as a single SDoc
+	  sdoc :: SDoc ErrAnnotation
+          sdoc    = formatErrDoc dflags errdoc
+
+	  -- we create a suitable context for producing calling 'collectAnns'
+	  sdocctx :: SDocContext
+	  sdocctx = initSDocContext dflags (defaultUserStyle dflags)
+
+Costs and Drawbacks
+-------------------
+
+The ``Outputable`` class in GHC lets us specify how to render values of all
+sorts of types as documents:::
+
+    class Outputable a where
```

Comment:
I still favor

```
class Outputable ty annot where
  ppr :: ty -> SDoc annot
```

Then, we can have

```
instance Outputable Type TcErrorAnnot where
  ppr = embed
instance {-# OVERLAPPABLE #-} Outputable Type a where
  ppr = pprType
```

Otherwise, we'll have to change a **lot** of calls to `ppr` throughout the codebase to be annotation-aware.

Reply:
Hmm, right, this does seem to be the nicest option as this will not incur additional work for people who emit those annotations. I guess we could even have an `Outputable Type Void` instance (and possibly others) in place of the generic, overlappable one -- in case someone objects to using overlappable instances for this.

Reply:
As far as I understand it, overlapping instances are bad only when combined with orphan instances. I think this one is pretty innocent. And it's useful: I might want `ppr :: Type -> SDoc ParseError` somehow.

Another approach: Conveniently, we have `fmap Data.Void.absurd :: forall a. SDoc Void -> SDoc a`.  We could say `fromVoidDoc = fmap absurd` and then if someone wants to print a `Type` in a `SDoc ParseError`, they could use the `Outputable Type Void` instance and `fromVoidDoc`. That's not so bad. I *think* it's valid to have `{-# RULES "fromVoidDoc" fromVoidDoc = unsafeCoerce #-}` to avoid rebuilding the whole `SDoc` tree just for a type change.

Maybe we present both the design with the overlapping instance and without and see what people prefer. Regardless, `fromVoidDoc` seems like a grand idea.

Reply:
Let's indeed mention the options we discussed here and see where the discussion goes.